instrumentation/graphql-java-12.0/library/src/main/java/io/opentelemetry/instrumentation/graphql/GraphqlTracingInstrumentation.java

```diff
@@ -0,0 +1,156 @@
+import graphql.execution.instrumentation.InstrumentationContext;
+import graphql.execution.instrumentation.InstrumentationState;
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.SimpleInstrumentationContext;
+import graphql.execution.instrumentation.parameters.InstrumentationExecuteOperationParameters;
+import graphql.execution.instrumentation.parameters.InstrumentationExecutionParameters;
+import graphql.language.AstPrinter;
+import graphql.language.AstTransformer;
+import graphql.language.BooleanValue;
+import graphql.language.EnumValue;
+import graphql.language.Node;
+import graphql.language.NodeVisitorStub;
+import graphql.language.NullValue;
+import graphql.language.OperationDefinition;
+import graphql.language.OperationDefinition.Operation;
+import graphql.language.Value;
+import graphql.language.VariableReference;
+import graphql.util.TraversalControl;
+import graphql.util.TraverserContext;
+import graphql.util.TreeTransformerUtil;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+
+public final class GraphqlTracingInstrumentation extends SimpleInstrumentation {
+  private final Instrumenter<InstrumentationExecutionParameters, ExecutionResult> instrumenter;
+  private final boolean captureExperimentalSpanAttributes;
+  private final boolean sanitizeQuery;
+
+  GraphqlTracingInstrumentation(
+      Instrumenter<InstrumentationExecutionParameters, ExecutionResult> instrumenter,
+      boolean captureExperimentalSpanAttributes,
+      boolean sanitizeQuery) {
+    this.instrumenter = instrumenter;
+    this.captureExperimentalSpanAttributes = captureExperimentalSpanAttributes;
+    this.sanitizeQuery = sanitizeQuery;
+  }
+
+  @Override
+  public InstrumentationState createState() {
+    return new GraphqlTracingInstrumentationState();
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginExecution(
+      InstrumentationExecutionParameters parameters) {
+
+    Context parentContext = Context.current();
+    if (!instrumenter.shouldStart(parentContext, parameters)) {
+      return SimpleInstrumentationContext.noOp();
+    }
+
+    Context context = instrumenter.start(parentContext, parameters);
+    GraphqlTracingInstrumentationState state = parameters.getInstrumentationState();
+    state.setContext(context);
+    Scope scope = context.makeCurrent();
+
+    return SimpleInstrumentationContext.whenCompleted(
+        (result, throwable) -> {
+          scope.close();
+
+          Span span = Span.fromContext(context);
+          for (GraphQLError error : result.getErrors()) {
+            AttributesBuilder attributes = Attributes.builder();
```

Comment:
Record errors (validation and parse errors etc) as exceptions. I don't know whether this is ok.

Reply:
I re-read the [exceptions spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md) and it seems ok. Is this what JS implementation does?

Reply:
As far as I understand they do something similar with validation errors https://github.com/open-telemetry/opentelemetry-js-contrib/blob/491588f7245e48268bea438cc475e87ae1ed8ad6/plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts#L385 but they don't seem to be doing anything with execution errors https://github.com/open-telemetry/opentelemetry-js-contrib/blob/491588f7245e48268bea438cc475e87ae1ed8ad6/plugins/node/opentelemetry-instrumentation-graphql/src/instrumentation.ts#L241